### PR TITLE
Fix issue #4399

### PR DESCRIPTION
**Pull request by pamod**
## Purpose
This resolves the issue described in https://github.com/ballerinalang/ballerina/issues/4399

## Goals
Avoid returning an additional string once the final record is processed. Fixes the issue mentioned in https://github.com/ballerinalang/ballerina/issues/4399

### CODE REVIEW

File: modules/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/io/channels/base/TextRecordChannel.java [hunk 1]
```diff
@@ -120,7 +120,7 @@ private String readRecord() throws BallerinaIOException {
             }
         } while (record == null && !readCharacters.isEmpty());
 
-        if (null == record && readCharacters.isEmpty()) {
+        if (null == record) {
```
Comment (hasithaa) [inline modules/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/io/channels/base/TextRecordChannel.java]: There is a code formatting issue here.

Comment (hasithaa) [reply modules/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/io/channels/base/TextRecordChannel.java]: Ohh My bad, It was aligned properly.

**Comment by pamod** [comment]
Record is a sequence of characters which is separated by a record separator (i.e character such as '/n'). Each record will have fields separated by a field separator.  

TextRecordChannel.read() API was designed to return an array of fields pertaining to a record. Per each read() the records will be read from the source (i.e file, TCP socket) iteratively. 

If all the records have being read from the channel an empty array will be returned. Indicating there're no records left.

In order to identify whether all the records have being read a flag (remaining) is maintained. 

Record reading flow in summary would be the following,

1. Check if there're remaining records to be read
2. If there're records the process will continue, if there aren't an empty array will be returned 
3. Read a defined number of characters from the Character I/O API 
4. Identify a record from the read character sequence, if there're excess characters read store them in memory (which will be leveraged when reading subsequent records).
5. while reading if there're no records left in memory and the I/O source returned EoF 
6. if EoF is reached and If there're characters left in memory (this will not form a complete record) these characters will be returned as the last record and remaining will be marked as false 
7. Once the record is identified the next step is to seperate the fields
6. The separated fields will be returned back to caller  

The issue  #4399 is product where the final records is read and in step 7 and there were no characters left in-memory. The record was returned as an empty string and the empty string was returned as the field 

Given the remaining flag is checked at step 1, the last record being empty and validation of whether the the channel has reached EoF was not checked at this stage. The provided fix will perform this validation.

**Comment by pamod** [comment]
@hasithaa test case added with 899c8bab18a7119aec9abd968f314e4ddf35f674

**Pull request merged**